plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkPublishOperation.java

```diff
@@ -0,0 +1,330 @@
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.core.runtime.IStatus;
+import org.eclipse.core.runtime.MultiStatus;
+import org.eclipse.core.runtime.Status;
+import org.eclipse.jst.server.core.IJ2EEModule;
+import org.eclipse.jst.server.core.IWebModule;
+import org.eclipse.wst.server.core.IModule;
+import org.eclipse.wst.server.core.IServer;
+import org.eclipse.wst.server.core.model.IModuleResource;
+import org.eclipse.wst.server.core.model.IModuleResourceDelta;
+import org.eclipse.wst.server.core.model.PublishOperation;
+import org.eclipse.wst.server.core.model.ServerBehaviourDelegate;
+import org.eclipse.wst.server.core.util.PublishHelper;
+
+import java.io.File;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Properties;
+
+public class CloudSdkPublishOperation extends PublishOperation {
+  private static final String PLUGIN_ID = CloudSdkPublishOperation.class.getName();
+
+  /**
+   * Throws new {@link CoreException} if status list is not empty.
+   */
+  private static void checkStatuses(List<IStatus> statusList) throws CoreException {
+    if (statusList == null || statusList.size() == 0) {
+      return;
+    }
+    if (statusList.size() == 1) {
+      throw new CoreException(statusList.get(0));
+    }
+    IStatus[] children = statusList.toArray(new IStatus[statusList.size()]);
+    throw new CoreException(
+        new MultiStatus(PLUGIN_ID, 0, children,
+        "Error during publish operation", null));
+  }
+
+  private CloudSdkServerBehaviour server;
+  private IModule[] modules;
+  private int kind;
+  private int deltaKind;
+  private PublishHelper helper;
+
+  public int getKind() {
+    return REQUIRED;
+  }
+
+  @Override
+  public int getOrder() {
+    // Hmm, shouldn't the publish tasks come later?
+    return 0;
+  }
+
+  /**
+   * Construct the operation object to publish the specified modules(s) to the specified server.
+   */
+  public CloudSdkPublishOperation(CloudSdkServerBehaviour server, int kind, IModule[] modules,
+      int deltaKind) {
+    super("Publish to server", "Publish a modules to App Engine Development Server");
+    this.server = server;
+    this.modules = modules;
+    this.kind = kind;
+    this.deltaKind = deltaKind;
+    IPath base = server.getRuntimeBaseDirectory();
+    helper = new PublishHelper(base.toFile());
+  }
+
+  @Override
+  public void execute(IProgressMonitor monitor, IAdaptable info) throws CoreException {
+    // TODO(amitin): use more advanced key to store modules publish locations? Because a dependent
+    // java project (added as child modules and published as jar) cannot present in more than one
+    // parent modules.
+    List<IStatus> statusList = Lists.newArrayList();
+    if (modules.length == 1) {
+      IPath deployPath = server.getModuleDeployDirectory(modules[0]);
+      // root modules
+      publishDir(deployPath, statusList, monitor);
+    } else {
+      // BSD: NEED TO UNDERSTAND SITUATION FOR REMAINING CODE
```

Comment:
For each module-set added to a server, we create a corresponding `CloudSdkPublishOperation` (done by `CloudSdkPublishTaskDelegate`).  I'm not clear when and why we might get multiple modules in a set. Perhaps .wars including other war projects?